libpromises/promises.c

```diff
@@ -240,12 +249,32 @@ Promise *DeRefCopyPromise(EvalContext *ctx, const Promise *pp)
 
                         if (IsDefinedClass(ctx, scp->classes))
                         {
+                            SeqAppend(constraints_to_expand, scp);
+                        }
+                    }
+
+                    // Only do the copy expansion if we copied something while inheriting
+                    // note that r < 0 can't be true because size_t is unsigned
+                    for (size_t r = SeqLength(constraints_to_expand); r != 0; r--)
+                    {
+                        Constraint *scp = SeqAt(constraints_to_expand, r-1);
+
+                        if (IsDefinedClass(ctx, scp->classes))
+                        {
                             Rval returnval = RvalNew(scp->rval.item, scp->rval.type);
 
                             // First we rewrite the Rval with the rewrite map
                             Rval rewrite = RvalCopyRewriter(returnval, arg_rewrite);
                             RvalDestroy(returnval);
 
+                            // TODO: this was only in the case of body call with no args, should it always happen?
```

Comment:
It's been that way since before the first Git import, different behavior between the two branches... I agree we need to commit to one way or the other, and this way is more logical.